srml/support/src/storage/storage_items.rs

```diff
@@ -832,6 +839,53 @@ mod test_map_vec_append {
 			assert_eq!(JustVec::get(), vec![1, 2, 3, 4, 5]);
 		});
 	}
+
+	#[test]
+	fn safe_append_works() {
+		with_externalities(&mut TestExternalities::default(), || {
+			let _ = MapVec::safe_append(1, &[1, 2, 3]);
+			let _ = MapVec::safe_append(1, &[4, 5]);
+			assert_eq!(MapVec::get(1), vec![1, 2, 3, 4, 5]);
+
+			let _ = JustVec::safe_append(&[1, 2, 3]);
+			let _ = JustVec::safe_append(&[4, 5]);
+			assert_eq!(JustVec::get(), vec![1, 2, 3, 4, 5]);
+		});
+	}
+
+	#[test]
+	fn len_works() {
+		with_externalities(&mut TestExternalities::default(), || {
+			JustVec::put(&vec![1, 2, 3, 4]);
+			OptionVec::put(&vec![1u32, 2, 3, 4, 5]);
+			MapVec::insert(1, &vec![1, 2, 3, 4, 5, 6]);
+
+			assert_eq!(JustVec::decode_len().unwrap(), 4);
+			assert_eq!(OptionVec::decode_len().unwrap(), 5);
+			assert_eq!(MapVec::decode_len(1).unwrap(), 6);
+		});
+	}
+
+	#[test]
+	fn len_works_for_uninitialized() {
+		with_externalities(&mut TestExternalities::default(), || {
+			assert_eq!(JustVec::get(), vec![]);
+			assert_noop!(JustVec::decode_len(), "could not find item to decode length");
+
+			assert_eq!(OptionVec::get(), None);
+			assert_noop!(OptionVec::decode_len(), "could not find item to decode length");
+
+			// NOTE: this must be WARNED and documented to the upmost extent.
+			assert_eq!(JustVecWithDefault::get(), vec![6u32, 9]);
```

Comment:
I just got an idea for this. We could introduce a marker trait. `NoDefault` or a similar name.

This trait is implemented by `decl_storage!` if there is no default given to `decl_storage!`. Then we can just add this trait bound for the `decode_len()` function.

Reply:
hmm good question actually I kind of think that `JustVecWithDefault::decode_len()` should return 2.
Also for append actually I kind of think that if the storage is empty it should append to default value instead of empty value.

Reply:
The traits are not aware of the default values.

Reply:
is this default feature still used ? at some point if it is not we could get rid of it and if a user really want to use it then it has to declare its own variable and implement trait Default on it with the value he wants.